packages/mongodb-access-adapter/datagrip-access-adapter/src/main/kotlin/com/mongodb/jbplugin/accessadapter/datagrip/adapter/DataGripMongoDbDriver.kt

```diff
@@ -33,60 +36,92 @@ import kotlinx.coroutines.withTimeout
+        .indent(false)
+        .build()
+
+    private fun Bson.toJson(): String = this.toBsonDocument().toJson(jsonWriterSettings)
+
+    override suspend fun connectionString(): ConnectionString = ConnectionString(dataSource.url!!)
 
     override suspend fun <T : Any> runCommand(
-        command: Document,
+        database: String,
+        command: Bson,
         result: KClass<T>,
-        timeout: Duration
-    ): T = withContext(
-        Dispatchers.IO
-    ) {
-        runQuery(
-            """db.runCommand(${command.toJson()})""",
-            result,
-            timeout
-        )[0]
-    }
+        timeout: Duration,
+    ): T =
+        withContext(
+            Dispatchers.IO,
+        ) {
+            runQuery(
+                """
+                    db.getSiblingDB("$database")
+                      .runCommand(EJSON.parse(`${command.toJson()}`))
+                """.trimIndent(),
+                result,
+                timeout,
+            )[0]
+        }
 
     override suspend fun <T : Any> findOne(
         namespace: Namespace,
-        query: Document,
-        options: Document,
+        query: Bson,
+        options: Bson,
         result: KClass<T>,
-        timeout: Duration
-    ): T? = withContext(Dispatchers.IO) {
-        runQuery(
-            """db.getSiblingDB("${namespace.database}")
+        timeout: Duration,
+    ): T? =
+        withContext(Dispatchers.IO) {
+            runQuery(
+                """db.getSiblingDB("${namespace.database}")
                  .getCollection("${namespace.collection}")
-                 .findOne(${query.toJson()}, ${options.toJson()}) """.trimMargin(),
-            result,
-            timeout
-        ).getOrNull(0)
-    }
+                 .findOne(EJSON.parse(`${query.toJson()}`), EJSON.parse(`${options.toJson()}`)) 
+                """.trimMargin(),
+                result,
+                timeout,
+            ).getOrNull(0)
+        }
 
     override suspend fun <T : Any> findAll(
         namespace: Namespace,
-        query: Document,
+        query: Bson,
         result: KClass<T>,
         limit: Int,
-        timeout: Duration
+        timeout: Duration,
     ) = withContext(Dispatchers.IO) {
         runQuery(
             """db.getSiblingDB("${namespace.database}")
                  .getCollection("${namespace.collection}")
-                 .find(${query.toJson()}).limit($limit) """.trimMargin(),
+                 .find(EJSON.parse(`${query.toJson()}`)).limit($limit) 
+            """.trimMargin(),
             result,
-            timeout
+            timeout,
         )
     }
 
+    override suspend fun countAll(
+        namespace: Namespace,
+        query: Bson,
+        timeout: Duration,
+    ) = withContext(Dispatchers.IO) {
+        runQuery(
+            """
+            db.getSiblingDB("${namespace.database}")
+                 .getCollection("${namespace.collection}")
+                 .countDocuments(EJSON.parse(`${query.toJson()}`))
```

Comment:
This is definitely beeter, but most queries here still have escaping issues (e.g.: the query could contain a literal `` ` ``, a database or collection name could include newlines, etc.)

Reply:
> The query is a JSON document which is escaped when using toJson, so it should be safe.

So `` `${query.toJson()}` `` is not actuall backticks surrounding the raw JSON String? Can you point me to where that escaping is happening?

> com.mongodb.jbplugin.accessadapter.Namespace.Companion#invoke

Okay, yeah, that makes sense, but ... it's honestly pretty far from obvious how this works. Would it be possible to name this so that `namespace.database` gives you the database name as a Java String and `namespace.escapedDatabase` gives you the escaped version? That is a bit more verbose but won't give every reader of this code a reason to scratch their heads 🙂

Reply:
That makes sense, I'm wondering how to do it in a way that is obvious which version to use, so we don't use the unescaped version in code.

Reply:
I changed the name of the properties to `escaped` and added the non escaped versions. To avoid misuse of the non escaped versions, I marked the API as delicate and needs to be opted in when used using an annotation so it won't compile.